postrprocess_visual/object_detection_visualizer.py

```diff
@@ -0,0 +1,217 @@
+import logging
+import os
+
+import cv2
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+import torchvision
+from torch.utils.data import Subset
+from torchvision.transforms import v2 as transforms
+
+from models.centernet import ModelBuilder
+from postrprocess_visual.postprocess import CenternetPostprocess
+from postrprocess_visual.visualizer import PASCAL_CLASSES
+from training.encoder import CenternetEncoder
+
+
+class ObjectDetectionVisualizer:
+    def __init__(
+        self,
+        input_height=256,
+        input_width=256,
+        down_ratio=4,
+        checkpoint_path=None,
+        confidence_threshold=0.3,
+    ):
+        logging.basicConfig(level=logging.INFO)
+        self.logger = logging.getLogger(__name__)
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.logger.info(f"Using device: {self.device}")
+
+        self.input_height = input_height
+        self.input_width = input_width
+        self.down_ratio = down_ratio
+        self.confidence_threshold = confidence_threshold
+
+        self.checkpoint_path = (
+            "../models/checkpoints/pretrained_weights.pt"
+            if checkpoint_path is None
+            else checkpoint_path
+        )
+
+        self._setup()
+
+    def _setup(self):
+        try:
+            self.dataset = self._prepare_dataset()
+            self.transform = self._create_transforms()
+            self.encoder = CenternetEncoder(self.input_height, self.input_width)
+            self.postprocessor = CenternetPostprocess(
+                n_classes=20,
+                width=self.input_width,
+                height=self.input_height,
+                down_ratio=self.down_ratio,
+            ).to(self.device)
+            self.model = self._load_trained_model()
+        except Exception as e:
+            self.logger.error(f"Error setting up components: {e}")
+            raise
+
+    def _prepare_dataset(self):
+        try:
+            dataset_val = torchvision.datasets.VOCDetection(
+                root="../VOC", year="2007", image_set="val", download=True
+            )
+            dataset_val = torchvision.datasets.wrap_dataset_for_transforms_v2(
+                dataset_val
+            )
+            indices = list(range(min(10, len(dataset_val))))
```

Comment:
`len(dataset_val)` is always > 10. Then this line can be reduced to `indices = range(10)`, as there is no need to convert indices to list.

Reply:
how this feature must be implemented, I propose make 2 modes
1) visualization on first 1...n images 
2) select 1...n random images from whole dataset

Reply:
done